// @fnando/streamdeck/dist/Encoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Encoder = void 0;
class Encoder {
    toManifest() {
        const snippet = {
            background: this.background,
            Icon: this.icon,
            layout: this.layout,
            StackColor: this.stackColor,
            TriggerDescription: {
                Rotate: this.onRotateDescription,
                Push: this.onPushDescription,
                Touch: this.onTouchDescription,
                LongTouch: this.onLongTouchDescription,
            },
        };
        const optionals = [
            ["background", this.background, this.background],
            ["Icon", this.icon, this.icon],
            ["layout", this.layout, this.layout],
            ["StackColor", this.stackColor, this.stackColor],
            ["StackColor", this.stackColor, this.stackColor],
            [
                "TriggerDescription",
                this.onRotateDescription ||
                    this.onPushDescription ||
                    this.onTouchDescription ||
                    this.onLongTouchDescription,
                this.buildTriggerDescription(),
            ],
        ];
        this.evaluateOptionalValues(optionals, snippet);
        return snippet;
    }
    buildTriggerDescription() {
        const snippet = {};
        const optionals = [
            ["Rotate", this.onRotateDescription, this.onRotateDescription],
            ["Push", this.onPushDescription, this.onPushDescription],
            ["Touch", this.onTouchDescription, this.onTouchDescription],
            ["LongTouch", this.onLongTouchDescription, this.onLongTouchDescription],
        ];
        this.evaluateOptionalValues(optionals, snippet);
        return snippet;
    }
    evaluateOptionalValues(optionals, object) {
        optionals.forEach(([prop, condition, value]) => {
            if (condition) {
                object[prop] = value;
            }
        });
        return object;
    }
}
exports.Encoder = Encoder;
