// @fnando/streamdeck/dist/State.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.State = void 0;
class State {
    constructor(params) {
        var _a, _b, _c, _d, _e;
        this.showTitle = true;
        this.name = "";
        this.title = "";
        this.image = "";
        this.multiActionImage = "";
        this.action = params.action;
        this.image = params.image;
        this.multiActionImage = (_a = params.multiActionImage) !== null && _a !== void 0 ? _a : "";
        this.showTitle = (_b = params.showTitle) !== null && _b !== void 0 ? _b : true;
        this.name = (_c = params.name) !== null && _c !== void 0 ? _c : "";
        this.title = (_d = params.title) !== null && _d !== void 0 ? _d : "";
        this.fontSize = (_e = params.fontSize) !== null && _e !== void 0 ? _e : 16;
        this.align = params.align;
        this.underline = params.underline;
        this.fontStyle = params.fontStyle;
        this.titleColor = params.titleColor;
    }
    toManifest() {
        const actionName = this.action.constructor.name;
        const baseImagePath = `images/actions/${actionName}/${this.image}`;
        const snippet = { Image: baseImagePath };
        const optionals = [
            [
                "MultiActionImage",
                this.multiActionImage,
                `src/images/actions/${actionName}/${this.multiActionImage}`,
            ],
            ["Name", this.name, this.name],
            ["Title", this.title, this.title],
            ["ShowTitle", this.showTitle === false, this.showTitle],
            ["FontSize", this.fontSize, String(this.fontSize)],
            ["FontUnderline", this.underline, this.underline],
            ["FontStyle", this.fontStyle, this.fontStyle],
            ["TitleAlignment", this.align, this.align],
            ["TitleColor", this.titleColor, this.titleColor],
        ];
        optionals.forEach(([prop, condition, value]) => {
            if (condition) {
                snippet[prop] = value;
            }
        });
        return snippet;
    }
}
exports.State = State;
